packages/autorest.go/src/generator/polymorphics.ts

```diff
@@ -3,137 +3,135 @@
+export async function generatePolymorphicHelpers(codeModel: GoCodeModel, packageName?: string): Promise<string> {
+  if (!codeModel.interfaceTypes) {
     // no polymorphic types
     return '';
   }
-  const discriminators = <Array<ObjectSchema>>session.model.language.go!.discriminators.filter((d: ObjectSchema) => !d.language.go!.omitType);
-  if (discriminators.length === 0) {
-    // all polymorphic types omitted
-    return '';
-  }
-  let text = await contentPreamble(session, packageName);
+
+  let text = contentPreamble(codeModel, packageName);
   const imports = new ImportManager();
   imports.add('encoding/json');
   if (packageName) {
     // content is being generated into a separate package, add the necessary import
-    imports.add(await getParentImport(session));
+    imports.add(getParentImport(codeModel));
   }
+
   text += imports.text();
-  // add any sub-hierarchies (SalmonType, SharkType in the test server) to the list
-  for (const disc of values(discriminators)) {
-    for (const val of values(disc.discriminator!.all)) {
-      const objSchema = <ObjectSchema>val;
-      // some hierarchies can overlap, so conditionally add
-      if (objSchema.discriminator && !discriminators.includes(objSchema)) {
-        discriminators.push(objSchema);
-      }
-    }
-  }
+
   const scalars = new Set<string>();
   const arrays = new Set<string>();
   const maps = new Set<string>();
-  const trackDisciminator = function(prop: Property) {
-    if (prop.schema.language.go!.discriminatorInterface) {
-      scalars.add(prop.schema.language.go!.discriminatorInterface);
-    } else if (isArraySchema(prop.schema)) {
-      const discriminatorInterface = recursiveUnwrapArrayDictionary(prop.schema).language.go!.discriminatorInterface;
-      if (discriminatorInterface) {
-        scalars.add(discriminatorInterface);
-        arrays.add(discriminatorInterface);
+
+  // we know there are polymorphic types but we don't know how they're used.
+  // i.e. are they vanilla fields, elements in a slice, or values in a map.
+  // polymorphic types within maps/slices will also need the scalar helpers.
+  const trackDisciminator = function(type: PossibleType) {
+    if (isInterfaceType(type)) {
+      scalars.add(type.name);
+    } else if (isSliceType(type)) {
+      const leafType = recursiveUnwrapMapSlice(type);
+      if (isInterfaceType(leafType)) {
+        scalars.add(leafType.name);
+        arrays.add(leafType.name);
       }
-    } else if (isDictionarySchema(prop.schema)) {
-      const discriminatorInterface = recursiveUnwrapArrayDictionary(prop.schema).language.go!.discriminatorInterface;
-      if (discriminatorInterface) {
-        scalars.add(discriminatorInterface);
-        maps.add(discriminatorInterface);
+    } else if (isMapType(type)) {
+      const leafType = recursiveUnwrapMapSlice(type);
+      if (isInterfaceType(leafType)) {
+        scalars.add(leafType.name);
+        maps.add(leafType.name);
       }
```

Comment:
What if its not one of the three types?  Is that ok?

This could be a case statement that throws if we the value is something other than an interface, slice or map.  Also are those mutually exclusive properties?